### PR TITLE
[ASTGen/Parse] Implement ASTGen changes for integer generics from Swift Syntax and restrict parsing integer types in certain contexts

### PR DESCRIPTION
**Pull request by Azoy**
In support of the Swift syntax parser being able to parse integer changes, we need to handle those in ASTGen.

Syntax PR: https://github.com/swiftlang/swift-syntax/pull/2859

### CODE REVIEW

**Comment by Azoy** [comment]
Please test with the following pull request:
https://github.com/swiftlang/swift-syntax/pull/2859

@swift-ci please test

**Comment by Azoy** [comment]
Please test with the following pull request:
https://github.com/swiftlang/swift-syntax/pull/2859

@swift-ci please test

**Comment by rintaro** [comment]
Please add tests in `lib/ASTGen/`.
Also, could we remove `-disable-experimental-parser-round-trip` in the following tests now?

test/Interpreter/value_generics.swift
test/ModuleInterface/value_generics.swift
test/Sema/value_generics.swift
test/Serialization/value_generics.swift

File: lib/ASTGen/Sources/ASTGen/Generics.swift [hunk 1]
```diff
@@ -87,4 +89,117 @@ extension ASTGenVisitor {
...
+  }
+
+  func generate(
+    integerType node: IntegerLiteralExprSyntax,
+    minusLoc: BridgedSourceLoc = BridgedSourceLoc()
+  ) -> BridgedTypeRepr {
+    var desc = node.trimmedDescription
+    let str = desc.withBridgedString {
+      self.ctx.allocateCopy(string: $0)
+    }
```
Comment (rintaro) [inline lib/ASTGen/Sources/ASTGen/Generics.swift:196]: Could you use `self.copyAndStripUnderscores(text: node.rawText)` (where `node` is `IntegerLiteralExprSyntax`)

File: lib/ASTGen/Sources/ASTGen/Generics.swift [hunk 2]
```diff
@@ -87,4 +89,117 @@ extension ASTGenVisitor {
...
+        return self.generate(integerType: node)
+
+      case .prefixOperatorExpr(let node):
+        let op = node.operator
+
+        guard op.text == "-" else {
+          fatalError("Unknown prefix operator for same type right type")
+        }
+
+        guard case .integerLiteralExpr(let node) = node.expression.as(ExprSyntaxEnum.self) else {
```
Comment (rintaro) [inline lib/ASTGen/Sources/ASTGen/Generics.swift:145]: Could be
```suggestion
        guard let node = node.expression.as(IntegerLiteralExprSyntax.self) else {
```

File: lib/ASTGen/Sources/ASTGen/Generics.swift [hunk 3]
```diff
@@ -87,4 +89,117 @@ extension ASTGenVisitor {
...
+      default:
+        fatalError("Unknown expression kind for generic argument type")
+      }
+    }
+  }
+
+  func generate(
+    integerType node: IntegerLiteralExprSyntax,
+    minusLoc: BridgedSourceLoc = BridgedSourceLoc()
+  ) -> BridgedTypeRepr {
```
Comment (rintaro) [inline lib/ASTGen/Sources/ASTGen/Generics.swift:192]:
```suggestion
  func generateIntegerType(
    integerLiteralExpr node: IntegerLiteralExprSyntax,
    minusLoc: BridgedSourceLoc = BridgedSourceLoc()
  ) -> IntegerLiteralTypeRepr {
```

ASTGen has a loose `generate` function naming convention `generate{Result}({source}:)` where `{Result}` is what to generate, and `{source}` is the input syntax type. Also we'd like to delay upcasting (`.asTypeRepr`) as much as possible.

Also could this function include the common logic between `case .expr(let expr):` bodies of the two functions above? i.e. 
```swift
  func generateIntegerType(expr node: Expr) -> BridgedIntegerTypeRepr {
    let minusLoc: SourceLoc
    let literalExpr: integerLiteralExprSyntax

    switch expr.as(ExprSyntaxEnum.self) {
    case .integerLiteralExpr(let node):
      minusLoc = nil
      literalExpr = node
    case .prefixOperatorExpr(let node):
      // .. check `-{IntergerLiteralSyntax}` ...
      minusLoc = self.generateSourceLoc(op)
      literalExpr = ...
    default:
        fatalError(...)
    }

    return .createParsed(
      self.ctx,
      string: self.copyAndStripUnderscores(text: literalExpr.rawText),
      loc: self.generateSourceLoc(literalExpr),
      minusLoc: minusLoc
    )
  }
```

**Review by rintaro — COMMENTED** [review]
lib/Parse and lib/ASTGen changes look good to me. Thanks!
Still missing `test/ASTGen` test cases. Also just a small issues in `SwiftMacros`

File: lib/Macros/Sources/SwiftMacros/DistributedResolvableMacro.swift [hunk 4]
```diff
@@ -141,10 +142,21 @@ extension DistributedResolvableMacro {
...
-           where sameTypeReq.leftType.isActorSystem:
-        specificActorSystemRequirement = sameTypeReq.rightType.trimmed
-        isGenericStub = false
+      case .sameTypeRequirement(let sameTypeReq):
+        switch sameTypeReq.leftType {
+        case .type(let type) where type.isActorSystem:
+          switch sameTypeReq.rightType.trimmed {
+          case .type(let rightType):
+            specificActorSystemRequirement = rightType
+            isGenericStub = false
```
[the remark this answers is not in the document]
Comment (Azoy) [reply lib/Macros/Sources/SwiftMacros/DistributedResolvableMacro.swift:151]: This is intended as there is another switch.

File: lib/Macros/Sources/SwiftMacros/DistributedResolvableMacro.swift [hunk 5]
```diff
@@ -141,10 +142,21 @@ extension DistributedResolvableMacro {
...
+
+          case .expr:
+            fatalError("Expression type not supported for distributed actor")
```
[the remark this answers is not in the document]
Comment (Azoy) [reply lib/Macros/Sources/SwiftMacros/DistributedResolvableMacro.swift:154]: I'll change this to throw a diagnostic instead.

**Comment by Azoy** [comment]
Please test with the following pull request:
https://github.com/swiftlang/swift-syntax/pull/2859
https://github.com/swiftlang/swift-format/pull/866

@swift-ci please test

File: lib/Macros/Sources/SwiftMacros/DistributedResolvableMacro.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (rintaro) [reply lib/Macros/Sources/SwiftMacros/DistributedResolvableMacro.swift:151]: Oh, my bad. I didn't see it

**Comment by Azoy** [comment]
Please test with the following pull request:
https://github.com/swiftlang/swift-syntax/pull/2859
https://github.com/swiftlang/swift-format/pull/866

@swift-ci please test

**Comment by Azoy** [comment]
Please test with the following pull request:
https://github.com/swiftlang/swift-syntax/pull/2859
https://github.com/swiftlang/swift-format/pull/866

@swift-ci please test

**Review by rintaro — APPROVED** [review]
Thank you for doing this!

**Comment by Azoy** [comment]
Please test with the following pull request:
https://github.com/swiftlang/swift-syntax/pull/2859
https://github.com/swiftlang/swift-format/pull/866

@swift-ci please test

**Pull request merged**